src/sentry/grouping/ingest/grouphash_metadata.py

```diff
@@ -71,8 +103,16 @@ def create_or_update_grouphash_metadata(
         grouphash.metadata.update(latest_grouping_config=grouping_config)
 
 
-def _get_hash_basis(event: Event, project: Project, variants: dict[str, BaseVariant]) -> HashBasis:
-    main_variant = (
+def get_hash_basis_and_metadata(
+    event: Event, project: Project, variants: dict[str, BaseVariant]
+) -> tuple[HashBasis, HashingMetadata]:
+    hashing_metadata: HashingMetadata = {}
+
+    # TODO: This (and `contributing_variant` below) are typed as `Any` so that we don't have to cast
+    # them to whatever specific subtypes of `BaseVariant` and `GroupingComponent` (respectively)
+    # each of the helper calls below requires. Casting once, to a type retrieved from a look-up,
```

Comment:
I believe the way to do this as written is to type narrow within the various if functions. e.g. on the first condition:

```
    if hash_basis == HashBasis.STACKTRACE:
        if not isinstance(
            contributing_component,
            StacktraceGroupingComponent
            | ExceptionGroupingComponent
            | ChainedExceptionGroupingComponent
            | ThreadsGroupingComponent,
        ):
            raise ValueError(
                "Contributing component is not a StacktraceGroupingComponent for stacktrace basis"
            )
        if not isinstance(contributing_variant, ComponentVariant):
            raise ValueError("Contributing variant is not a ComponentVariant for stacktrace basis")

        hashing_metadata = _get_stacktrace_hashing_metadata(
            contributing_variant, contributing_component
        )
```

I would argue for replacing [conditional with polymorphism](https://refactoring.guru/replace-conditional-with-polymorphism) as an alternative, but understand that given existing grouping code, this way is easier.

Reply:
Oh, yeah... something like this, probably?

```python
class BaseHashBasis[
    VariantType: BaseVariant,
    ComponentType: BaseGroupingComponent[Any] | None,
    MetadataType: HashingMetadata,
]:
    name: str
    variants: dict[str, BaseVariant]
    contributing_variant: VariantType
    contributing_component: ComponentType | None

    def __init__(
        self,
        variants: dict[str, BaseVariant],
        contributing_variant: VariantType,
        contributing_component: ComponentType | None = None,
    ):
        self.variants = variants
        self.contributing_variant = contributing_variant
        self.contributing_component = contributing_component

        if contributing_component:
            self.subcomponents_by_id = {
                subcomponent.id: subcomponent for subcomponent in contributing_component.values
            }

    def get_hashing_metadata(self) -> HashingMetadata:
        raise NotImplementedError()


class TemplateHashBasis(
    BaseHashBasis[ComponentVariant, TemplateGroupingComponent, TemplateHashingMetadata]
):
    name: str = "template"

    def get_hashing_metadata(self):
        metadata: TemplateHashingMetadata = {}

        if self.subcomponents_by_id["filename"].values:
            metadata["template_name"] = self.subcomponents_by_id["filename"].values[0]
        if self.subcomponents_by_id["context-line"].values:
            metadata["template_context_line"] = self.subcomponents_by_id["context-line"].values[0]

        return metadata

def get_hash_basis_and_metadata(
    event: Event, project: Project, variants: dict[str, BaseVariant]
) -> tuple[HashBasis, HashingMetadata]:
    # Skipping some of the details below, but you get the idea
    contributing_variant = ...
    contributing_component = ...

    hash_basis, hash_basis_type = GROUPING_METHODS_BY_DESCRIPTION[method_description]

    hashing_metadata = hash_basis_type(
        variants, contributing_variant, contributing_component
    ).get_hashing_metadata()

    return hash_basis, hashing_metadata 
```

Would the typechecker be any happier with that, though? Specifically, wouldn't I still have to type `contributing_variant` and `contributing_component` as `Any` in order to pass them to the `TemplateHashBasis`/etc constructor?

Reply:
I was thinking moreso putting formatting methods on the variant / component classes, and aiming to remove hash-basis as its own enum if possible. i don't have time right now to go into details, so please feel free to ignore, or look at the type narrowing using `isinstance`. whatever's easiest and can help get this merged faster.